# Exploring eBay Car Sales Data

The aim of this project is to clean the dataset and perform some initial analysis on it.

## Opening and exploring data:

Data dictionary for the dataset:

* **dateCrawled** - When the ad was first crawled. All field-values are taken from this date.

* **name** - Name of the car.

* **seller** - Whether the seller is private or a dealer.

* **offerType** - The type of listing.

* **price** - The listed selling price of the car.

* **abtest** - Whether the listing is included in an A/B test.

* **vehicleType** - The type of vehicle.

* **yearOfRegistration** - The year in which the car was first registered.

* **gearbox** - The type of transmission.

* **powerPS** - The power of the car in PS.

* **model** - The car model name.

* **odometer** - How many kilometers the car has driven.

* **monthOfRegistration** - The month in which the car was first registered.

* **fuelType** - What type of fuel the car uses.

* **brand** - The brand of the car.

* **notRepairedDamage** - If the car has a damage which is not yet repaired.

* **dateCreated** - The date the eBay listing was created.

* **nrOfPictures** - The number of pictures in the ad.

* **postalCode** - The postal code for the location of the vehicle.

* **lastSeenOnline** - When the crawler saw this ad last online. this ad last online.

In [5]:
import numpy as np
import pandas as pd
autos=pd.read_csv('autos.csv',encoding = 'latin1')

To make them easier to explore, we created a function named autos.info() and autos.head(5).

In [7]:
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column               Non-Null Count   Dtype 
---  ------               --------------   ----- 
 0   dateCrawled          371528 non-null  object
 1   name                 371528 non-null  object
 2   seller               371528 non-null  object
 3   offerType            371528 non-null  object
 4   price                371528 non-null  int64 
 5   abtest               371528 non-null  object
 6   vehicleType          333659 non-null  object
 7   yearOfRegistration   371528 non-null  int64 
 8   gearbox              351319 non-null  object
 9   powerPS              371528 non-null  int64 
 10  model                351044 non-null  object
 11  kilometer            371528 non-null  int64 
 12  monthOfRegistration  371528 non-null  int64 
 13  fuelType             338142 non-null  object
 14  brand                371528 non-null  object
 15  notRepairedDamage    299468 non-nu

In [8]:
autos.head(5)

dateCrawled                            name  seller offerType  \
0  2016-03-24 11:52:17                      Golf_3_1.6  privat   Angebot   
1  2016-03-24 10:58:45            A5_Sportback_2.7_Tdi  privat   Angebot   
2  2016-03-14 12:52:21  Jeep_Grand_Cherokee_"Overland"  privat   Angebot   
3  2016-03-17 16:54:04              GOLF_4_1_4__3TÜRER  privat   Angebot   
4  2016-03-31 17:25:20  Skoda_Fabia_1.4_TDI_PD_Classic  privat   Angebot   

   price abtest vehicleType  yearOfRegistration    gearbox  powerPS  model  \
0    480   test         NaN                1993    manuell        0   golf   
1  18300   test       coupe                2011    manuell      190    NaN   
2   9800   test         suv                2004  automatik      163  grand   
3   1500   test  kleinwagen                2001    manuell       75   golf   
4   3600   test  kleinwagen                2008    manuell       69  fabia   

   kilometer  monthOfRegistration fuelType       brand notRepairedDamage  \
0     150000                    0   benzin  volkswagen               NaN   
1     125000                    5   diesel        audi                ja   
2     125000                    8   diesel        jeep               NaN   
3     150000                    6   benzin  volkswagen              nein   
4      90000                    7   diesel       skoda              nein   

           dateCreated  nrOfPictures  postalCode             lastSeen  
0  2016-03-24 00:00:00             0       70435  2016-04-07 03:16:57  
1  2016-03-24 00:00:00             0       66954  2016-04-07 01:46:50  
2  2016-03-14 00:00:00             0       90480  2016-04-05 12:47:46  
3  2016-03-17 00:00:00             0       91074  2016-03-17 17:40:17  
4  2016-03-31 00:00:00             0       60437  2016-04-06 10:17:21

A [dataset](https://data.world/data-society/used-cars-data) containing data about approximately 371528 cars for sale:
* The dataset contains 20 columns, most of which are strings.
* Some columns are missed data such as: 'vehicleType', 'gearbox', 'model', 'fuelType','notRepairedDamage'.
* The column names use [camelcase](https://en.wikipedia.org/wiki/Camel_case) instead of Python's preferred [snakecase](https://en.wikipedia.org/wiki/Snake_case), we will replace spaces with underscores.

## Cleaning Column Names

Let's convert the column names from camelcase to snakecase and change some of the column names based on the data dictionary in order to be more descriptive.

In [12]:
autos.columns

Index(['dateCrawled', 'name', 'seller', 'offerType', 'price', 'abtest',
       'vehicleType', 'yearOfRegistration', 'gearbox', 'powerPS', 'model',
       'kilometer', 'monthOfRegistration', 'fuelType', 'brand',
       'notRepairedDamage', 'dateCreated', 'nrOfPictures', 'postalCode',
       'lastSeen'],
      dtype='object')

In [13]:
def change_name(dataset, old_names, new_names):
    rename_dict = dict(zip(old_names, new_names))
    dataset = dataset.rename(columns=rename_dict)
    return dataset

In [14]:
    old_names = ["yearOfRegistration", "monthOfRegistration", "notRepairedDamage", "dateCreated"]
    new_names = ["registration_year", "registration_month", "unrepaired_damage", "ad_created"]
    
    autos = change_name(autos, old_names, new_names)

In [15]:
import re

def change_case(dataset):
    new_columns = []
    for col in dataset.columns:
        col = re.sub(r'([a-z])([A-Z])', r'\1_\2', col).lower()
        new_columns.append(col)
    dataset.columns = new_columns
    return dataset


In [16]:
autos = change_case(autos)
autos.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 20 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   date_crawled        371528 non-null  object
 1   name                371528 non-null  object
 2   seller              371528 non-null  object
 3   offer_type          371528 non-null  object
 4   price               371528 non-null  int64 
 5   abtest              371528 non-null  object
 6   vehicle_type        333659 non-null  object
 7   registration_year   371528 non-null  int64 
 8   gearbox             351319 non-null  object
 9   power_ps            371528 non-null  int64 
 10  model               351044 non-null  object
 11  kilometer           371528 non-null  int64 
 12  registration_month  371528 non-null  int64 
 13  fuel_type           338142 non-null  object
 14  brand               371528 non-null  object
 15  unrepaired_damage   299468 non-null  object
 16  ad

## Initial Exploration

Now let's do some basic data exploration to determine what other cleaning tasks need to be done. Initially we looked for:
1. Some columns that have mostly one value that are candidates to be dropped: 'seller', 'offer_type', 'abtest'
2. Some columns that have no value for analysis that are candidates to be dropped : 'date_crawled', 'ad_created', 'nr_of_pictures', 'postal_code','last_seen'
3. Some columns need more investigation:
    * The have a minimum value of 0 is 'price','power_ps'                
    * The 'power_ps' column and the 'price' has outliers           
    * The 'registration_year' column contains some odd values  The minimum value is 1000, before cars were invened.  The maximum value is 9999, many years into the future

autos.tail(5)

In [20]:
autos.describe(include='all')

date_crawled         name  seller offer_type         price  \
count                371528       371528  371528     371528  3.715280e+05   
unique               280500       233531       2          2           NaN   
top     2016-03-24 14:49:47  Ford_Fiesta  privat    Angebot           NaN   
freq                      7          657  371525     371516           NaN   
mean                    NaN          NaN     NaN        NaN  1.729514e+04   
std                     NaN          NaN     NaN        NaN  3.587954e+06   
min                     NaN          NaN     NaN        NaN  0.000000e+00   
25%                     NaN          NaN     NaN        NaN  1.150000e+03   
50%                     NaN          NaN     NaN        NaN  2.950000e+03   
75%                     NaN          NaN     NaN        NaN  7.200000e+03   
max                     NaN          NaN     NaN        NaN  2.147484e+09   

        abtest vehicle_type  registration_year  gearbox       power_ps  \
count   371528       333659      371528.000000   351319  371528.000000   
unique       2            8                NaN        2            NaN   
top       test    limousine                NaN  manuell            NaN   
freq    192585        95894                NaN   274214            NaN   
mean       NaN          NaN        2004.577997      NaN     115.549477   
std        NaN          NaN          92.866598      NaN     192.139578   
min        NaN          NaN        1000.000000      NaN       0.000000   
25%        NaN          NaN        1999.000000      NaN      70.000000   
50%        NaN          NaN        2003.000000      NaN     105.000000   
75%        NaN          NaN        2008.000000      NaN     150.000000   
max        NaN          NaN        9999.000000      NaN   20000.000000   

         model      kilometer  registration_month fuel_type       brand  \
count   351044  371528.000000       371528.000000    338142      371528   
unique     251            NaN                 NaN         7          40   
top       golf            NaN                 NaN    benzin  volkswagen   
freq     30070            NaN                 NaN    223857       79640   
mean       NaN  125618.688228            5.734445       NaN         NaN   
std        NaN   40112.337051            3.712412       NaN         NaN   
min        NaN    5000.000000            0.000000       NaN         NaN   
25%        NaN  125000.000000            3.000000       NaN         NaN   
50%        NaN  150000.000000            6.000000       NaN         NaN   
75%        NaN  150000.000000            9.000000       NaN         NaN   
max        NaN  150000.000000           12.000000       NaN         NaN   

       unrepaired_damage           ad_created  nr_of_pictures   postal_code  \
count             299468               371528        371528.0  371528.00000   
unique                 2                  114             NaN           NaN   
top                 nein  2016-04-03 00:00:00             NaN           NaN   
freq              263182                14450             NaN           NaN   
mean                 NaN                  NaN             0.0   50820.66764   
std                  NaN                  NaN             0.0   25799.08247   
min                  NaN                  NaN             0.0    1067.00000   
25%                  NaN                  NaN             0.0   30459.00000   
50%                  NaN                  NaN             0.0   49610.00000   
75%                  NaN                  NaN             0.0   71546.00000   
max                  NaN                  NaN             0.0   99998.00000   

                  last_seen  
count                371528  
unique               182806  
top     2016-04-07 06:45:59  
freq                     17  
mean                    NaN  
std                     NaN  
min                     NaN  
25%                     NaN  
50%                     NaN  
75%                     NaN  
max                  

## Cleaning

### Drop some columns

In [24]:
autos.columns

Index(['date_crawled', 'name', 'seller', 'offer_type', 'price', 'abtest',
       'vehicle_type', 'registration_year', 'gearbox', 'power_ps', 'model',
       'kilometer', 'registration_month', 'fuel_type', 'brand',
       'unrepaired_damage', 'ad_created', 'nr_of_pictures', 'postal_code',
       'last_seen'],
      dtype='object')

In [25]:
df=autos[['name', 'price', 'vehicle_type', 'registration_year', 'gearbox', 'power_ps',  
             'model','kilometer', 'registration_month', 'fuel_type', 'brand','unrepaired_damage']]

In [26]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371528 entries, 0 to 371527
Data columns (total 12 columns):
 #   Column              Non-Null Count   Dtype 
---  ------              --------------   ----- 
 0   name                371528 non-null  object
 1   price               371528 non-null  int64 
 2   vehicle_type        333659 non-null  object
 3   registration_year   371528 non-null  int64 
 4   gearbox             351319 non-null  object
 5   power_ps            371528 non-null  int64 
 6   model               351044 non-null  object
 7   kilometer           371528 non-null  int64 
 8   registration_month  371528 non-null  int64 
 9   fuel_type           338142 non-null  object
 10  brand               371528 non-null  object
 11  unrepaired_damage   299468 non-null  object
dtypes: int64(5), object(7)
memory usage: 34.0+ MB


### Removed the outliers

 We'll start by analyzing the 'price'columns

In [29]:
print(df["price"].unique().shape[0])
print('\n')
print(df["price"].describe())
print('\n')
print(df["price"].value_counts().sort_index(ascending= True).head(10))
print('\n')
print(df["price"].value_counts().sort_index(ascending= False).head(10))

5597


count    3.715280e+05
mean     1.729514e+04
std      3.587954e+06
min      0.000000e+00
25%      1.150000e+03
50%      2.950000e+03
75%      7.200000e+03
max      2.147484e+09
Name: price, dtype: float64


price
0     10778
1      1189
2        12
3         8
4         1
5        26
7         3
8         9
9         8
10       84
Name: count, dtype: int64


price
2147483647     1
99999999      15
99000000       1
74185296       1
32545461       1
27322222       1
14000500       1
12345678       9
11111111      10
10010011       1
Name: count, dtype: int64


In [30]:
df=df[df["price"].between(500,20000)]

In [31]:
print(df["price"].describe())

count    318322.000000
mean       4895.139243
std        4520.389885
min         500.000000
25%        1499.000000
50%        3200.000000
75%        6950.000000
max       20000.000000
Name: price, dtype: float64


We focus on cars have price from 500 to 10000

Next, We'll start by analyzing the 'power_ps'columns

In [34]:
print(df["power_ps"].unique().shape[0])
print('\n')
print(df["power_ps"].describe())
print('\n')
print(df["power_ps"].value_counts().sort_index(ascending= True).head(10))
print('\n')
print(df["power_ps"].value_counts().sort_index(ascending= False).head(10))

668


count    318322.000000
mean        114.945637
std         189.172572
min           0.000000
25%          75.000000
50%         107.000000
75%         145.000000
max       20000.000000
Name: power_ps, dtype: float64


power_ps
0    28666
1       17
2        9
3        9
4       28
5       98
6       10
7       10
8        7
9        4
Name: count, dtype: int64


power_ps
20000    1
19312    1
19211    1
19208    1
17932    1
17700    1
17410    1
17019    1
17011    1
16312    1
Name: count, dtype: int64


We focus on cars have power_ps from 20 to 300

In [36]:
df=df[df["power_ps"].between(20,300)]
print(df["power_ps"].describe())

count    285966.000000
mean        120.338659
std          48.563845
min          20.000000
25%          80.000000
50%         115.000000
75%         150.000000
max         300.000000
Name: power_ps, dtype: float64


Next, We'll start by analyzing the 'registration_year'columns

In [38]:
df["registration_year"].describe()

count    285966.000000
mean       2003.744260
std          29.543278
min        1000.000000
25%        2000.000000
50%        2004.000000
75%        2008.000000
max        9999.000000
Name: registration_year, dtype: float64

In [40]:
df=df[df["registration_year"].between(1900,2024)]
print(df["registration_year"].describe())

count    285946.000000
mean       2003.570520
std           6.836097
min        1910.000000
25%        2000.000000
50%        2004.000000
75%        2008.000000
max        2019.000000
Name: registration_year, dtype: float64


In [41]:
print(df.isnull().sum())

name                      0
price                     0
vehicle_type          18108
registration_year         0
gearbox                5086
power_ps                  0
model                 10502
kilometer                 0
registration_month        0
fuel_type             16518
brand                     0
unrepaired_damage     40248
dtype: int64


In [42]:
print(df['vehicle_type'].value_counts(dropna=False))

vehicle_type
limousine     78711
kleinwagen    62405
kombi         56545
bus           25840
cabrio        18261
NaN           18108
coupe         13409
suv           10457
andere         2210
Name: count, dtype: int64


In [43]:
print(df['gearbox'].value_counts(dropna=False))

gearbox
manuell      224225
automatik     56635
NaN            5086
Name: count, dtype: int64


In [44]:
print(df['model'].value_counts(dropna=False))

model
golf                  24307
andere                19184
3er                   17101
NaN                   10502
polo                   9670
                      ...  
i3                        4
samara                    3
rangerover                3
serie_3                   3
range_rover_evoque        2
Name: count, Length: 250, dtype: int64


In [45]:
print(df['fuel_type'].value_counts(dropna=False))

fuel_type
benzin     175864
diesel      88350
NaN         16518
lpg          4416
cng           482
hybrid        202
andere         75
elektro        39
Name: count, dtype: int64


In [46]:
print(df['unrepaired_damage'].value_counts(dropna=False))

unrepaired_damage
nein    222192
NaN      40248
ja       23506
Name: count, dtype: int64


In [47]:
df=df.dropna()

In [48]:
df.head()

name  price vehicle_type  \
3                                  GOLF_4_1_4__3TÜRER   1500   kleinwagen   
4                      Skoda_Fabia_1.4_TDI_PD_Classic   3600   kleinwagen   
5   BMW_316i___e36_Limousine___Bastlerfahrzeug__Ex...    650    limousine   
6                         Peugeot_206_CC_110_Platinum   2200       cabrio   
10                                  Mazda_3_1.6_Sport   2000    limousine   

    registration_year  gearbox  power_ps    model  kilometer  \
3                2001  manuell        75     golf     150000   
4                2008  manuell        69    fabia      90000   
5                1995  manuell       102      3er     150000   
6                2004  manuell       109  2_reihe     150000   
10               2004  manuell       105  3_reihe     150000   

    registration_month fuel_type       brand unrepaired_damage  
3                    6    benzin  volkswagen              nein  
4                    7    diesel       skoda              nein  
5                   10    benzin         bmw                ja  
6                    8    benzin     peugeot              nein  
10                  12    benzin       mazda              nein

## Exploring Price, Mileage by Brand

When working with data on cars, it's natural to explore variations across different car brands. We can use aggregation to understand the brand column.

In [51]:
top_20_brand=df["brand"].value_counts().head(20).index
print("Top 20 favourite car:", top_20_brand)

Top 20 favourite car: Index(['volkswagen', 'bmw', 'opel', 'mercedes_benz', 'audi', 'ford', 'renault',
       'peugeot', 'fiat', 'seat', 'skoda', 'mazda', 'toyota', 'citroen',
       'smart', 'nissan', 'mini', 'hyundai', 'volvo', 'mitsubishi'],
      dtype='object', name='brand')


In [52]:
brand_mean_price={}
for b in top_20_brand:
    mean_price=df.loc[df["brand"]==b,"price"].mean()
    brand_mean_price[b]=mean_price
bmp_series = pd.Series(brand_mean_price).head(6)
print(bmp_series)

volkswagen       5497.725710
bmw              7008.477106
opel             3601.726392
mercedes_benz    6539.648714
audi             7261.766218
ford             4150.165980
dtype: float64


In [53]:
brand_mean_mileage={}
for b in top_20_brand:
    mean_mileage=df.loc[df["brand"]==b,"kilometer"].mean()
    brand_mean_mileage[b]=mean_mileage
bmm_series = pd.Series(brand_mean_mileage).head(6)
print(bmm_series)

volkswagen       128557.902251
bmw              137393.805658
opel             126307.567591
mercedes_benz    135188.934803
audi             135369.379283
ford             122733.378333
dtype: float64


In [59]:
df_mean = pd.DataFrame(bmp_series, columns=['mean_price'])
df_mean['mean_mileage']=bmm_series
df_mean

mean_price   mean_mileage
volkswagen     5497.725710  128557.902251
bmw            7008.477106  137393.805658
opel           3601.726392  126307.567591
mercedes_benz  6539.648714  135188.934803
audi           7261.766218  135369.379283
ford           4150.165980  122733.378333

The average number of brand's mileage is slightly differently

Luxury brands such as Audi, BMW, and Mercedes-Benz were consistently priced higher, even if their mileage was not low. This aligns with the market positioning of these brands.

We observed that generally, higher prices do not seem to correlate with mileage. For instance, luxury brands maintained high prices regardless of their mileage.